registry/v2/registry.go

```diff
@@ -0,0 +1,257 @@
+	URL        *url.URL
+	tlsConfig  *tls.Config
+	httpClient *http.Client
+	Username   string
+	Password   string
+}
+
+type Repo struct {
+	Namespace  string
+	Repository string
+}
+
+type TagList struct {
+	Tags []string `json:"tags"`
+}
+
+func newHTTPClient(u *url.URL, tlsConfig *tls.Config, timeout time.Duration) *http.Client {
+	httpTransport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	httpTransport.Dial = func(proto, addr string) (net.Conn, error) {
+		return net.DialTimeout(proto, addr, timeout)
+	}
+	return &http.Client{Transport: httpTransport}
+}
+
+func NewRegistryClient(registryUrl string, tlsConfig *tls.Config, username string, password string) (*RegistryClient, error) {
+	u, err := url.Parse(registryUrl)
+	if err != nil {
+		return nil, err
+	}
+	httpClient := newHTTPClient(u, tlsConfig, defaultHTTPTimeout)
+	return &RegistryClient{
+		URL:        u,
+		httpClient: httpClient,
+		tlsConfig:  tlsConfig,
+		Username:   username,
+		Password:   password,
+	}, nil
+}
+
+func (client *RegistryClient) doRequest(method string, path string, body []byte, headers map[string]string) ([]byte, http.Header, error) {
+	b := bytes.NewBuffer(body)
+
+	req, err := http.NewRequest(method, client.URL.String()+"/v2"+path, b)
+	log.Debugf("Method: %s   URL: %s", method, client.URL.String()+"/v2"+path)
+	if err != nil {
+		log.Debugf("Error on doRequest")
+		return nil, nil, err
+	}
+
+	req.SetBasicAuth(client.Username, client.Password)
+
+	req.Header.Add("Content-Type", "application/json")
+	if headers != nil {
+		for header, value := range headers {
+			req.Header.Add(header, value)
+		}
+	}
+
+	resp, err := client.httpClient.Do(req)
+	if err != nil {
+		if !strings.Contains(err.Error(), "connection refused") && client.tlsConfig == nil {
+			return nil, nil, fmt.Errorf("%s. Are you trying to connect to a TLS-enabled daemon without TLS?", err)
+		}
+		log.Debugf("Connection refused")
+		return nil, nil, err
+	}
+
+	defer resp.Body.Close()
+
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Debugf("Error on ioutil.ReadAll")
+		return nil, nil, err
+	}
+
+	if resp.StatusCode == 404 {
+		log.Debugf("Error on resp.StatusCode == 404")
+		return nil, nil, ErrNotFound
+	}
+
+	if resp.StatusCode >= 400 {
+		log.Debugf("Error on resp.StatusCode >= 400")
+		return nil, nil, Error{StatusCode: resp.StatusCode, Status: resp.Status, msg: string(data)}
+	}
+
+	return data, resp.Header, nil
+}
+
+func (client *RegistryClient) Search(query string) ([]*Repository, error) {
+	type repo struct {
+		Repositories []string `json:"repositories"`
+	}
+
+	uri := fmt.Sprintf("/_catalog")
+	data, _, err := client.doRequest("GET", uri, nil, nil)
+	if err != nil {
+		log.Debugf("Error on client.doRequest(GET, uri, nil, nil)")
+		return nil, err
+	}
+
+	log.Debugf("Response from /_catalog %s", string(data))
+
+	res := &repo{}
+	if err := json.Unmarshal(data, &res); err != nil {
+		log.Debugf("Error on json.Unmarshal(data, &res)")
```

Comment:
remove or more descriptive (if more descriptive fix the `Debugf` -- you aren't passing anything to format)